Sources/Plasma/NucleusLib/pnAsyncCore/Private/pnAcThread.h

```diff
@@ -54,6 +54,33 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 
 #include "pnNetBase/pnNbError.h"
 #include "pnUtils/pnUtils.h"
+#include "hsThread.h"
+
+template<class _Fp, class ..._Args>
+class pnAcThread: public std::thread {
+public:
+    explicit pnAcThread(_Fp&& __f, _Args&&... __args) :
+        fFuncPtr(__f),
+        std::thread(this, __args...) {
+    };
+    bool TimedJoin(unsigned timeoutMs) {
+        if (fSem.Wait(std::chrono::milliseconds(timeoutMs))) {
+            join();
+            return true;
+        } else {
+            detach();
+            return false;
+        }
+    }
+    
+private:
+    void operator()(_Args&&... args) {
+        fFuncPtr(args...);
+        fSem.Signal();
+    }
+    _Fp&& fFuncPtr;
+    hsSemaphore fSem = hsSemaphore();
+};
```

Comment:
Proposing something here - haven't wired it in yet.

This is a subclass of std::thread that implements a timed join. It has a few benefits:
- Just needs a semaphore to function. Semaphore is signaled when the function completes - so no need to rely on functionality in the underlying threading library. And no breaking std::thread by calling underlying function.
- Implements the same basic thread functionality we're using on Windows, and works everywhere.
- Properly implements timeout or join, no risk of a join firing later and messing things up if the thread was detached.

Downside: We'd need to use this subclass instead of std::thread where we need to use TimedJoin.

Reply:
Trying this out in practice. This implementation won't be sufficient in since in implementation threads are copied. That also means mutex ownership needs to move. Still working through that.

I guess I'm also starting to wonder what we're trying to accomplish here as well. The timeout based join will prevent deadlocks, but may also cause bad behavior by terminating threads mid execution. I'd be worried it we're relying on that timeout.